EchoPro/survey.py

```diff
@@ -562,98 +593,237 @@ def impute_missing_sigma_bs(self):
+
+        Notes
+        -----
+        This function uses the previously fitted length-weight relationship that, 
+        alongside the measured male-female proportions with predicted weight values,
+        is used to calculate the mean weight for males, females, and total animals 
+        for each strata.
+        """ 
+        
+        ### Reformat 'specimen_df' to match the same format as 'len_df'
+        # First make copies of each
+        specimen_df_copy = self.biology['specimen_df'].copy().pipe( lambda df: df.loc[ df.species_id == species_id ] )
+        length_df_copy = self.biology['length_df'].copy().pipe( lambda df: df.loc[ df.species_id == species_id ] )
+        
+        ### Pull length distribution values
+        length_intervals = self.biology['distributions']['length']['length_interval_arr']   
+        
+        ### Calculate the sex proportions/frequencies for station 1 (length_df) across all strata        
+        length_grouped = (   
+            length_df_copy
+            .discretize_variable( length_intervals , 'length' ) # appends `length_bin` column
+            .assign( group = lambda x: np.where( x['sex'] == int(1) , 'male' , 'female' ) ) # assigns str variable for comprehension
+            .pipe( lambda df: pd.concat( [ df.loc[ df[ 'sex' ] != 3 ] , df.assign( group = 'all' ) ] ) ) # appends male-female to an 'all' dataframe
+            .assign( station = 1 ) # assign station number for later functions
+            )
+        
+        ### Calculate the sex proportions/frequencies for station 2 (specimen_df) across all strata
+        specimen_grouped = (
+            specimen_df_copy
+            .discretize_variable( length_intervals , 'length' ) # appends `length_bin` column
+            .assign( group = lambda x: np.where( x['sex'] == int(1) , 'male' , 'female' ) ) # assigns str variable for comprehension
+            .pipe( lambda df: pd.concat( [ df.loc[ df[ 'sex' ] != 3 ] , df.assign( group = 'all' ) ] ) ) # appends male-female to an 'all' dataframe
+            .assign( station = 2 ) # assign station number for later functions
+        )
+        
+        ### "Meld" the two datasets together to keep downstream code tidy
+        station_sex_length = (
+            specimen_grouped # begin reformatting specimen_grouped so it resembles same structure as length_grouped
+            .meld( length_grouped ) # "meld": reformats specimen_grouped and then concatenates length_grouped 
+            # sum up all of the length values with respect to each length bin
+            .count_variable( contrasts = [ 'group' , 'station' , 'stratum_num' , 'length_bin' ] , # grouping variables
+                             variable = 'length_count' , # target value to apply a function
+                             fun = 'sum' ) # function to apply
+        )
+        
+        ### Calculate total sample size ('group' == 'all') that will convert 'counts' into 'frequency/proportion'
```

Comment:
The discrepancy between the original code and how this is summed here. This seems to be how the stratum sample size was calculated; the original code expressly omits values where `sex = 3` from the `specimen` dataset. There was a TODO comment included in there that just used the full sum of both the `length` and `specimen` dataframes for that calculation. 

For the most part, these differences were non-existence where `sex=3` was a trivial fraction of the data. When `stratum=2`, this is where the largest differences are/were observed, which coincides with a non-trivial number of unsexed animals where both the male and female samples comprise ~ 25% of the overall sample (i.e. ~ 50% of the sample from both datasets were sexed). So while the proportions for both `male` and `female` are identical between the original and updated code, the mean weights are slightly different.